components/datetime/src/pattern/item/ule.rs

```diff
@@ -0,0 +1,217 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::PatternItem;
+use crate::fields;
+use core::convert::TryFrom;
+use zerovec::ule::{AsULE, ULE};
+
+/// `PatternItemULE` is a type optimized for efficent storing and
+/// deserialization of `DateTimeFormat` `PatternItem` elements using
+/// `ZeroVec` model.
+///
+/// The serialization model packages the pattern item in three bytes.
+///
+/// The first bit is used to disriminate the item variant. If the bit is
+/// set, then the value is the `PatternItem::Field` variant. Otherwise,
+/// the `PatternItem::Literal` is used.
+///
+/// In case the discriminant is set:
+///
+/// 1) The rest of the first byte remains usnused.
+/// 2) The second byte encodes `FieldSymbol` encoded as (Type: 4 bits, Symbol: 4 bits).
+/// 3) The third byte encodes the field length.
+///
+/// If the discriminant is not set, the bottom three bits of the first byte,
+/// together with the next two bytes, contain all 21 bits required to encode
+/// any [`Unicode Code Point`].
+///
+/// # Diagram
+///
+/// <pre>
+/// ┌───────────────┬───────────────┬───────────────┐
+/// │       u8      │       u8      │       u8      │
+/// ├─┬─┬─┬─┬─┬─┬─┬─┼─┬─┬─┬─┬─┬─┬─┬─┼─┬─┬─┬─┬─┬─┬─┬─┤
+/// ├─┴─┴─┼─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┤
+/// │     │          Unicode Code Point             │ Literal
+/// ├─┬───┴─────────┬───────────────┬───────────────┤
+/// │X│             │  FieldSymbol  │  FieldLength  │ Field
+/// └─┴─────────────┴───────────────┴───────────────┘
+///  ▲
+///  │
+///  Variant Discriminant
+/// </pre>
+///
+/// # Optimization
+///
+/// This model is optimized for efficient packaging of the `PatternItem` elements
+/// and performant deserialization from the `PatternItemULE` to `PatternItem` type.
+///
+/// # Constraints
+///
+/// The model leaves at most 8 `PatternItem` variants, limits the number of possible
+/// field types and symbols to 16 each and limits the number of length variants to 256.
+///
+/// [`Unicode Code Point`]: http://www.unicode.org/versions/latest/
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct PatternItemULE([u8; 3]);
+
+impl PatternItemULE {
+    #[inline]
+    fn determine_field_from_u8(byte: u8) -> bool {
```

Comment:
issue: please document what is being determined here, since "field" is somewhat ambiguous